graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java

```diff
@@ -45,17 +58,30 @@ public AppConfigResource(Configuration configuration) {
         } catch (IOException e) {
             throw new RuntimeException("Unable to read AppConfig template while generating web interface configuration: ", e);
         }
+
+        final String serverTransportUri = serverUriFromHeadersOrDefault(headers);
         final Map<String, Object> model = new HashMap<String, Object>() {{
             put("rootTimeZone", configuration.getRootTimeZone());
-            put("serverUri", configuration.getRestTransportUri());
+            put("serverUri", serverTransportUri);
             put("appPathPrefix", "");
         }};
-        this.content = engine.transform(template, model);
+        return engine.transform(template, model);
     }
 
-    @GET
-    @Produces(MediaType.APPLICATION_OCTET_STREAM)
-    public String get() {
-        return this.content;
+    private String serverUriFromHeadersOrDefault(HttpHeaders httpHeaders) {
+        final List<String> overrideServerUriHeaders = httpHeaders.getRequestHeader(CK_OVERRIDE_SERVER_URI);
+        if (overrideServerUriHeaders != null && !overrideServerUriHeaders.isEmpty()) {
+            Optional<String> firstHeader = overrideServerUriHeaders.stream().filter(s -> {
+                try {
+                    return s != null && Strings.isNotEmpty(s) && new URL(s) != null;
```

Comment:
Unfortunately `URL` will execute a DNS lookup which (since this is called every time the `/config.js` resource is being requested) might slow down the system as a whole. Also see http://www.eishay.com/2008/04/javas-url-little-secret.html

Please use [`URI`](https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#URI(java.lang.String)) instead.